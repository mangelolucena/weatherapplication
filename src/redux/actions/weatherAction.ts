export enum ActionType {
    GET_WEATHER_REQUEST = "GET_WEATHER_REQUEST",
    GET_WEATHER_SUCCESS = "GET_WEATHER_SUCCESS",
    GET_WEATHER_FAILURE = "GET_WEATHER_FAILURE",
}

export const getWeatherRequest = (params: any) => ({
    type: ActionType.GET_WEATHER_REQUEST,
    payload: params
});

export const getWeatherSuccess = (params: any) => ({
    type: ActionType.GET_WEATHER_SUCCESS,
    payload: params
});

export const getWeatherFailure = (params: any) => ({
    type: ActionType.GET_WEATHER_FAILURE,
    payload: params
});
