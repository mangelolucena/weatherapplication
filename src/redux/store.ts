import { createStore, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from '@redux-saga/core';
import combineReducer from '../redux/reducers/index';
import { watcherSaga } from './sagas/weatherSaga';

const sagaMiddleware = createSagaMiddleware();

let store = createStore(
  combineReducer,
  compose(applyMiddleware(sagaMiddleware))
);
sagaMiddleware.run(watcherSaga);


export default store;
