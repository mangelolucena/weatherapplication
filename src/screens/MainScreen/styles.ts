import {StyleSheet} from 'react-native';
import {base} from 'styles';

export const styles = StyleSheet.create({
  container: {
    flex: base.container.flex,
  },
});
