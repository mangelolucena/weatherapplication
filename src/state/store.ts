import { createStore, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from '@redux-saga/core';
import combineReducer from './combineReducers';
import { watcherSaga } from './weather/weatherSaga';

const sagaMiddleware = createSagaMiddleware();

let store = createStore(
  combineReducer,
  compose(applyMiddleware(sagaMiddleware))
);
sagaMiddleware.run(watcherSaga);


export default store;
