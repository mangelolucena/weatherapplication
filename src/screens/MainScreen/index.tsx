import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { View, ScrollView, Button } from "react-native";
import { map } from 'lodash';
import moment from 'moment'

import { TextMedium } from "../../components/TextComponent";
import { ContainerComponent } from '../../components/ContainerComponent/index';
import { ImageComponent } from "../../components/ImageComponent";
import { CardComponent } from '../../components/CardComponent/index';
import { TextCustomSize, TextExtraLarge } from '../../components/TextComponent/index';

import { getWeatherRequest } from '../../redux/actions/weatherAction';

import { styles } from "./styles";

function getWeatherIcon(iconName: string) {
  return `http://openweathermap.org/img/wn/${iconName}@4x.png`;
}

function renderWeatherDetails(weather: any) {
  const weatherItems = [
    {
      label: 'Description',
      value: weather.weather[0].description
    },
    {
      label: 'Feels like',
      value: `${weather.main.feels_like} 째C`
    },
    {
      label: 'Humidity',
      value: `${weather.main.humidity} %`
    },
    {
      label: 'Pressure',
      value: `${weather.main.pressure} hPa`
    },
    {
      label: 'Clouds',
      value: `${weather.clouds.all} %`
    },
    {
      label: 'Wind',
      value: `${weather.wind.speed} meter/sec`
    },
    {
      label: 'Sunrise',
      value: moment.unix(weather.sys.sunrise).format('h:mm a')
    },
    {
      label: 'Sunset',
      value: moment.unix(weather.sys.sunset).format('h:mm a')
    }
  ];
  return (
    map(weatherItems, (items, i) =>
      <CardComponent key={i} style={styles.weatherDetailItems}>
        <TextMedium bold={true}>{`${items.label}: `}</TextMedium>
        <TextMedium>{items.value}</TextMedium>
      </CardComponent>
    )
  )
}


export default function MainScreen() {
  const dispatch = useDispatch();
  const state = useSelector((state: any) => state.weatherReducer)
  const [sampleStatus, setSampleStatus] = useState('');

  useEffect(() => {
    setTimeout(() => {
      setSampleStatus('timeout is called')
    }, 1000)
  }, [])

  useEffect(() => {
    // getCurrentLocation().then((location: any) => {
    //     this.props.getWeatherRequest({
    //         lat: location.latitude,
    //         lon: location.longitude,
    //         units: 'metric'
    //     });
    // });
    dispatch(getWeatherRequest({
      q: 'Pasig',
      units: 'metric'
    }))
  }, [state.weather]);
  return (
    <ContainerComponent>
      <Button testID='myButton' title='Sample' onPress={() => setSampleStatus('press')} />
      <TextExtraLarge testID='myText'>{sampleStatus}</TextExtraLarge>
      {
        state.weather ?
          <View style={styles.container}>
            <CardComponent style={styles.weatherMainContainer} >
              <View style={styles.rowContainer}>
                <View>

                  <View style={styles.weatherIconContainer}>
                    <ImageComponent source={{ uri: getWeatherIcon(state.weather.weather[0].icon) }} />
                  </View>
                </View>
                <View style={styles.weatherMainDataContainer}>
                  <TextExtraLarge testID={'weatherMain'} bold={true}>{state.weather.weather[0].main}</TextExtraLarge>
                  <TextCustomSize bold={true} fontSize={40}>{`${state.weather.main.temp} 째C`}</TextCustomSize>
                  <TextMedium>{`${state.weather.main.temp_max} 째C | ${state.weather.main.temp_min} 째C`}</TextMedium>
                </View>
              </View>
              <TextMedium style={styles.date}>{moment().format('MMMM D, h:mm a')}</TextMedium>
            </CardComponent>
            <View>
              <ScrollView>{renderWeatherDetails(state.weather)}</ScrollView>
            </View>
          </View >
          : null
      }
    </ContainerComponent >
  )
}