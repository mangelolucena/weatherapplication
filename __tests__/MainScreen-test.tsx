/**
 * @format
 */

import 'react-native';
import React from 'react';
import MainScreen from '../src/screens/MainScreen';
import Axios from 'axios';

// Note: test renderer must be required after react-native.
import { create, act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../src/redux/store';
import { getWeatherRequest } from '../src/redux/actions/weatherAction';

// jest.useFakeTimers()
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
const tree =
  create(
    <Provider store={store}>
      <MainScreen />
    </Provider>,
  )
test('renders main screen correctly', () => {
  act(() => {
    store.dispatch(getWeatherRequest({
      q: 'Pasig',
      units: 'metric'
    }))
  })

  tree.toJSON();
  expect(tree).toMatchSnapshot();
});


//handle state
test('button press', () => {
  const button = tree.root.findByProps({ testID: 'myButton' }).props;
  act(() => button.onPress())
  const text = tree.root.findByProps({ testID: 'myText' }).props;
  expect(text.children).toEqual('press')
})

//async task
test('call timeout', () => {
  act(() => {
    jest.runAllTimers()
  });
  const text = tree.root.findByProps({ testID: 'myText' }).props;
  expect(text.children).toEqual('timeout is called')
})

//redux
test('status stored properly', () => {
  expect(store.getState().weatherReducer.weather).toBeTruthy
})

//axios
test('request', () => {
  Axios.request.mockImplementation(() =>
    Promise.resolve({ data: { status: 'request is called' } }))
  act(() => { jest.runAllTimers() });

  const text = tree.root.findByProps({ testID: 'myText' }).props;
  expect(text.children).toEqual('timeout is called')
})

