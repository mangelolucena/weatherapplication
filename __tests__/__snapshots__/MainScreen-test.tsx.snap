// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders main screen correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <BVLinearGradient
    colors={
      Array [
        4278236891,
        4287745990,
        4292342511,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={
      Array [
        0.5,
        0.7,
        0.9,
      ]
    }
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      Object {
        "flex": 1,
        "padding": 20,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="myButton"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Sample
        </Text>
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#242424",
          "fontSize": 25,
          "fontWeight": "normal",
        }
      }
      testID="myText"
    >
      
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "white",
              "borderRadius": 8,
              "elevation": 5,
              "shadowColor": "gray",
              "shadowOffset": Object {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.25,
              "shadowRadius": 3.84,
            },
            Object {
              "alignItems": "center",
              "flex": 1,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <View
                accessibilityIgnoresInvertColors={true}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                <Image
                  onLoad={[Function]}
                  resizeMode="contain"
                  source={
                    Object {
                      "uri": "http://openweathermap.org/img/wn/04d@4x.png",
                    }
                  }
                  style={
                    Object {
                      "bottom": 0,
                      "height": 150,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                      "width": 150,
                    }
                  }
                  testID="RNE__Image"
                  theme={
                    Object {
                      "colors": Object {
                        "black": "#242424",
                        "disabled": "hsl(208, 8%, 90%)",
                        "divider": "#bcbbc1",
                        "error": "#ff190c",
                        "grey0": "#393e42",
                        "grey1": "#43484d",
                        "grey2": "#5e6977",
                        "grey3": "#86939e",
                        "grey4": "#bdc6cf",
                        "grey5": "#e1e8ee",
                        "greyOutline": "#bbb",
                        "platform": Object {
                          "android": Object {
                            "error": "#f44336",
                            "grey": "rgba(0, 0, 0, 0.54)",
                            "primary": "#2196f3",
                            "secondary": "#9C27B0",
                            "success": "#4caf50",
                            "warning": "#ffeb3b",
                          },
                          "ios": Object {
                            "error": "#ff3b30",
                            "grey": "#7d7d7d",
                            "primary": "#007aff",
                            "searchBg": "#dcdce1",
                            "secondary": "#5856d6",
                            "success": "#4cd964",
                            "warning": "#ffcc00",
                          },
                          "web": Object {
                            "error": "#ff190c",
                            "grey": "#393e42",
                            "primary": "#2089dc",
                            "searchBg": "#303337",
                            "secondary": "#ca71eb",
                            "success": "#52c41a",
                            "warning": "#faad14",
                          },
                        },
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                        "white": "#ffffff",
                      },
                    }
                  }
                  transition={true}
                  transitionDuration={360}
                />
                <View
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  style={
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "opacity": 1,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#bdbdbd",
                        "height": 150,
                        "justifyContent": "center",
                        "width": 150,
                      }
                    }
                    testID="RNE__Image__placeholder"
                  />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "height": 150,
                        "width": 150,
                      },
                      undefined,
                    ]
                  }
                />
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              bold={true}
              style={
                Object {
                  "color": "#242424",
                  "fontSize": 25,
                  "fontWeight": "bold",
                }
              }
              testID="weatherMain"
            >
              Clouds
            </Text>
            <Text
              bold={true}
              fontSize={40}
              style={
                Object {
                  "color": "#242424",
                  "fontSize": 40,
                  "fontWeight": "bold",
                }
              }
            >
              303.2 °C
            </Text>
            <Text
              style={
                Object {
                  "color": "#242424",
                  "fontSize": 15,
                  "fontWeight": "normal",
                }
              }
            >
              304.26 °C | 302.15 °C
            </Text>
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#242424",
              "fontSize": 15,
              "fontWeight": "normal",
              "marginTop": 10,
              "position": "absolute",
            }
          }
        >
          March 5, 7:55 pm
        </Text>
      </View>
      <View>
        <RCTScrollView>
          <View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 8,
                    "elevation": 5,
                    "shadowColor": "gray",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.25,
                    "shadowRadius": 3.84,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginTop": 8,
                    "padding": 20,
                  },
                ]
              }
            >
              <Text
                bold={true}
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "bold",
                  }
                }
              >
                Description: 
              </Text>
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "normal",
                  }
                }
              >
                broken clouds
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 8,
                    "elevation": 5,
                    "shadowColor": "gray",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.25,
                    "shadowRadius": 3.84,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginTop": 8,
                    "padding": 20,
                  },
                ]
              }
            >
              <Text
                bold={true}
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "bold",
                  }
                }
              >
                Feels like: 
              </Text>
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "normal",
                  }
                }
              >
                306.06 °C
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 8,
                    "elevation": 5,
                    "shadowColor": "gray",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.25,
                    "shadowRadius": 3.84,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginTop": 8,
                    "padding": 20,
                  },
                ]
              }
            >
              <Text
                bold={true}
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "bold",
                  }
                }
              >
                Humidity: 
              </Text>
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "normal",
                  }
                }
              >
                62 %
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 8,
                    "elevation": 5,
                    "shadowColor": "gray",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.25,
                    "shadowRadius": 3.84,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginTop": 8,
                    "padding": 20,
                  },
                ]
              }
            >
              <Text
                bold={true}
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "bold",
                  }
                }
              >
                Pressure: 
              </Text>
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "normal",
                  }
                }
              >
                1011 hPa
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 8,
                    "elevation": 5,
                    "shadowColor": "gray",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.25,
                    "shadowRadius": 3.84,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginTop": 8,
                    "padding": 20,
                  },
                ]
              }
            >
              <Text
                bold={true}
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "bold",
                  }
                }
              >
                Clouds: 
              </Text>
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "normal",
                  }
                }
              >
                75 %
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 8,
                    "elevation": 5,
                    "shadowColor": "gray",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.25,
                    "shadowRadius": 3.84,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginTop": 8,
                    "padding": 20,
                  },
                ]
              }
            >
              <Text
                bold={true}
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "bold",
                  }
                }
              >
                Wind: 
              </Text>
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "normal",
                  }
                }
              >
                2.6 meter/sec
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 8,
                    "elevation": 5,
                    "shadowColor": "gray",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.25,
                    "shadowRadius": 3.84,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginTop": 8,
                    "padding": 20,
                  },
                ]
              }
            >
              <Text
                bold={true}
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "bold",
                  }
                }
              >
                Sunrise: 
              </Text>
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "normal",
                  }
                }
              >
                6:21 am
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 8,
                    "elevation": 5,
                    "shadowColor": "gray",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.25,
                    "shadowRadius": 3.84,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginTop": 8,
                    "padding": 20,
                  },
                ]
              }
            >
              <Text
                bold={true}
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "bold",
                  }
                }
              >
                Sunset: 
              </Text>
              <Text
                style={
                  Object {
                    "color": "#242424",
                    "fontSize": 15,
                    "fontWeight": "normal",
                  }
                }
              >
                5:37 pm
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </BVLinearGradient>
</RCTSafeAreaView>
`;
